This is just a quick intro to using Julia in `jupyter notebook`. This follows the lessons by Dr. Lorena Barba which can be found at [AeroPython](https://github.com/barbagroup/AeroPython).

## Libraries

You can use many different packages for plotting in Julia (you can even use matplotlib).  The main way to add packages are to start the Julia power shell, press **]** key and the use the command "**add Package_Name**." The other way to add packages is shown in the following cell. We will plot using the regular Julia package **Plots** so we will add that package.

In [1]:
using Pkg

In [ ]:
Pkg.add("Plots")

Once the package is added we can tell Julia that we want to use it by issuing the **using Package_name** command in our script. We can then use the plot commands in our script. To suppress any output in Julia, we use a **;** at the end of a line as in Matlab. This is also true for plots. When using a jupyter notebook, the plot will show up as an output as long as the **;** is not present.

In [3]:
using Plots

Whereas matlab and numpy use the command [`linspace()`](http://docs.scipy.org/doc/numpy/reference/generated/numpy.linspace.html) to create an array, Julia uses a similar command in [range()](https://en.wikibooks.org/wiki/Introducing_Julia/Arrays_and_tuples). Note that you can also instantiate an array just as in matlab but using **a = start:step:end**, but range allows you to specify a length of your array and automatically calculates the step size.

In [4]:
myarray = range(0, stop=5, length=10);
println(myarray)

0.0:0.5555555555555556:5.0


Note that instead of printing out the whole array, it prints the start:step:end syntax. Lets try to use that sintax instead of range and just omit the **;** so that it is output instead of printing it:

In [5]:
myarray = 0.0:0.56:5.0

0.0:0.56:4.48

You can see that it doesn't quite get to 5.0. If we took 4.48 and added 0.56, we would get 5.04 which would be above 5.0. So this syntax steps at the required step length until it gets as close to the end number without going over it.

## Variables

Julia will infer the type of a variable by how you declare it. An interesting thing about Julua is that the name of a variable can be any unicode name. This means that you can even use backslashed LaTeX symbols to name your variables as shown below.

In [6]:
a = 5;       # a is an the integer 5
σ = "five";   # \sigma is a string of the word 'five'
c = 5.0;      # c is a floating point 5

We can see what type Julia infered from the variables by using the **typeof()** command.

In [7]:
typeof(a)

Int64

In [8]:
typeof(σ)

String

In [9]:
typeof(c)

Float64

We can divide integers, floats, or a combination that outputs a float with the correct value as below:

In [10]:
14/a

2.8

In [11]:
14/c

2.8

As with python, we get the same floating point number.

## For Loops

For loops in Julia work very much like a mix of matlab and python. You can use iterators as in python, but you need to use the range and put an end after the for loop as in matlab. 

In [12]:
for i in 1:5
    print("Hi \n")
end

Hi 
Hi 
Hi 
Hi 
Hi 


Nested for loops work the same way with each end defining the end of a loop. It's good practice to properly indent so that it's easy to read.

In [13]:
for i in 1:3
   for j in 1:3
        println(i, j)
    end
    print("This statement marks the end of the inner loop")
end

11
12
13
This statement marks the end of the inner loop21
22
23
This statement marks the end of the inner loop31
32
33
This statement marks the end of the inner loop

## Slicing Arrays

Slicing arrays means that you want to take a portion or subset of that array.  This can be either 1 value or a range of values.

In [14]:
myvals = [1, 2, 3, 4, 5]

5-element Array{Int64,1}:
 1
 2
 3
 4
 5

Julia uses a 1-based index like Matlab. Lets look at the first and last element:

In [15]:
myvals[1], myvals[5]

(1, 5)

Note that you could also use the special **end** value to get the end of the array:

In [16]:
myvals[end]

5

In order to get a subset of array, we use the syntax **start:end** within the index.

In [17]:
myvals[1:3]

3-element Array{Int64,1}:
 1
 2
 3

Like Python, if we define a variable equal to another variable, it is pointing to that variables location in memory. This means that any changes to the original array will also be changed in the other array that is pointing to the original array. Below is an example where we define an array variable equal to an array **a** and then change **a** and we see that it also changes **b**.

In [18]:
a = [1,2,3,4,5]

5-element Array{Int64,1}:
 1
 2
 3
 4
 5

In [19]:
b = a
println(b)

[1, 2, 3, 4, 5]


In [20]:
a[3] = 17
println(a)
println(b)

[1, 2, 17, 4, 5]
[1, 2, 17, 4, 5]


To make a proper copy of an array, we want to use the method **copy** as follows:

In [21]:
c = copy(a);

and now we change the variable in a back to 3 and check all the arrays we've made so far.

In [22]:
a[3] = 3
println("a = ", a)
println("b = ", b)
println("c = ", c)

a = [1, 2, 3, 4, 5]
b = [1, 2, 3, 4, 5]
c = [1, 2, 17, 4, 5]


We see that both **a** and **b** have their third element go back to 3 but the third element in **c** has stayed the same.